components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,561 @@
+            }
+            ParseOperation::CreateComment(text, id) => {
+                let comment = Comment::new(DOMString::from(text), document);
+                assert!(self.nodes.borrow_mut().insert(id, JS::from_ref(&comment.upcast())).is_none());
+            }
+            ParseOperation::Insert(parent, reference_child, new_node) => {
+                let nodes = self.nodes.borrow();
+                let parent = nodes.get(&parent).unwrap();
+                let reference_child = reference_child.map(|n| nodes.get(&n).unwrap().deref());
+
+                let node = match new_node {
+                    ParserNodeOrText::Node(ref id) => {
+                        let deref_node = nodes.get(id).unwrap().deref();
+                        NodeOrText::AppendNode(JS::from_ref(deref_node))
+                    },
+                    ParserNodeOrText::Text(ref string) => {
+                        match Tendril::from_str(string) {
+                            Ok(tendril) => NodeOrText::AppendText(tendril),
+                            Err(_) => panic!("Error while converting from string to tendril!"),
+                        }
+                    }
+                };
+
+                super::insert(parent, reference_child, node);
+            }
+            ParseOperation::AppendDoctypeToDocument(name, public_id, system_id) => {
+                let doctype = DocumentType::new(
+                    DOMString::from(name), Some(DOMString::from(public_id)),
+                    Some(DOMString::from(system_id)), document);
+
+                document.upcast::<Node>().AppendChild(doctype.upcast()).expect("Appending failed");
+            }
+            ParseOperation::AddAttrsIfMissing(target_id, attrs) => {
+                match self.nodes.borrow().get(&target_id) {
+                    Some(target) => {
+                        let elem = target.downcast::<Element>()
+                            .expect("tried to set attrs on non-Element in HTML parsing");
+                        for attr in attrs {
+                            elem.set_attribute_from_parser(attr.name, DOMString::from(String::from(attr.value)), None);
+                        }
+                    },
+                    None => panic!("Node not found!")
+                }
+            }
+            ParseOperation::RemoveFromParent(target) => {
+                match self.nodes.borrow().get(&target) {
+                    Some(target) => {
+                        if let Some(ref parent) = target.GetParentNode() {
+                            parent.RemoveChild(&*target).unwrap();
+                        }
+                    }
+                    None => panic!("Node not found!")
+                }
+            }
+            ParseOperation::MarkScriptAlreadyStarted(node) => {
+                match self.nodes.borrow().get(&node) {
+                    Some(node) => {
+                        let script = node.downcast::<HTMLScriptElement>();
```

Comment:
Use `expect` here, this operation should be done only with `HTMLScriptElement` values anyway.